### PR TITLE
[Aio] Run IO operations in a separated Asyncio loop

### PR DESCRIPTION
**Pull request by pfreixes**
All IO operations are executed using its own loop and thread, allowing the synchronous client to run without getting blocked.

Recipe for running the synchronous tests using the `Aio` iomgr:

```bash
python ./src/python/grpcio_tests/setup.py test_sync_over_aio
```
Related to #19955

## Performance impact

The following table shows the performance of the Python gRPC driver in its three different ways of making RPC calls comparing the PR with master. Each value shows the maximum QPS reached with the same level of concurrency (25 threads/25 Asyncio tasks)

_stack_ | Master | PR
--- | --- | ---
sync | 11K | 11K
sync over Aio | x | 6K
Aio| 19K| 18K

For testing has been using the following tool [1] for stressing a CPP gRPC server from a Python gRPC client by making unary calls, having both of them living in different machines. All tests that required the `Aio` module were executed using the `uvloop` implementation of the  Asyncio loop.

[1] https://github.com/pfreixes/grpcio_stress

## What is missing

- The `init_grpc_aio` is still the way of initializing the `Aio` module.
- Several tests have been disabled [1] we would need to tackle the ones that are critical.
- The new test command for running all of the synchronous tests on top of an initialized `Aio` module would need to be added into the CI for having them running automatically at each new PR.
- For detecting regressions with the solution proposed we could also consider having an entry into the current benchmark tests for checking the performance. We could consider testing all arities

## Other notes

None last minor versions of 3.6 and 3.7 have this bug [2] where asynchronous generators might end up being closed from another thread, all last minor versions have this bug fixed.

[1] https://github.com/grpc/grpc/pull/22076/files#diff-f4932b2c9567e7075b28bc8ddf591120R266
[2] https://bugs.python.org/issue34769

---

**EDIT(@lidizheng)**: This PR needs cherry-pick import. Fixes https://github.com/grpc/grpc/issues/19985

### CODE REVIEW

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        except Exception as exp:
+            print("An error ocurred with the IO loop {}".format(exp))
+            # Without the IO loop running the program would become
+            # unresponsive, proactively we close the process.
+            sys.exit(1)
+
+    cpdef void io_mark(self):
+        # Wake up all threads that were waiting
+        # for an IO event.
+        self._io_ev.set()
```
Comment (gnossen) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: I'm not sure I understand the usage of this `threading.Event`. Shouldn't calling `notify_all()` on a `threading.Condition` have the same effect? In fact, that's[ what `theading.Event` seems to be doing under the hood anyway](https://github.com/python/cpython/blob/f2ee21d858bc03dd801b97afe60ee2ea289e2fe9/Lib/threading.py#L565).

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        self._loop_started = False
+
+        self._thread = threading.Thread(target=self._run_forever, daemon=True)
+        self._thread.start()
+
+        # Some attributes are initialitzated when
+        # the thread is really started, we wait till
+        # the whole Asyncio loop is initialized and
+        # ready to be used.
+        with self._loop_started_cv:
```
Comment (gnossen) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: If this used a `threading.Event` instead, this could just be `self._loop_started_event.wait()`.

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            if not self._loop_started:
+                self._loop_started_cv.wait()
+
+        _io_loop = self
+
+
+    def _loop_started_cb(self):
+        with self._loop_started_cv:
```
Comment (gnossen) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: Likewise, this could just be `self._loop_started_event.set()`.

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        # Wake up all threads that were waiting
+        # for an IO event.
+        self._io_ev.set()
+
+        # Clear the status, further threads will get
+        # block.
+        self._io_ev.clear()
+
+    cdef void io_wait(self, size_t timeout_ms):
+        if threading.get_ident() == self._thread.ident:
```
Comment (gnossen) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: This needs to be a more explicit error condition. We need to bubble an error or at the very least log one.

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        self._asyncio_loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(self._asyncio_loop)
+        self._asyncio_loop.call_soon(self._loop_started_cb)
+        try:
+            self._asyncio_loop.run_forever()
+        except Exception as exp:
+            print("An error ocurred with the IO loop {}".format(exp))
+            # Without the IO loop running the program would become
+            # unresponsive, proactively we close the process.
+            sys.exit(1)
```
Comment (gnossen) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: When the event loop ran on the main thread, people had the ability to catch exceptions that bubbled up and handle them in an application-appropriate manner. Have we lost a significant capability by moving to a different thread?

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/iomgr.pyx.pxi [hunk 6]
```diff
@@ -208,10 +205,13 @@ cdef void asyncio_kick_loop() with gil:
 
 
 cdef void asyncio_run_loop(size_t timeout_ms) with gil:
-    pass
-
+    # Block until arrives a new input/output event
+    _current_io_loop().io_wait(timeout_ms)
```
Comment (lidizheng) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/iomgr.pyx.pxi]: 👍

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+    def __init__(self):
+        global _io_loop
+
+        # Only one instantiation is exepected        
+        assert _io_loop is None
+
+        self._asyncio_loop = None
+        self._io_ev = threading.Event()
+        self._loop_started_cv = threading.Condition()
```
Comment (lidizheng) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: Maybe we can use a `threading.Event()` here as well?

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/socket.pyx.pxi [hunk 8]
```diff
@@ -106,25 +111,35 @@ cdef class _AsyncioSocket:
...
-        self._task_connect.add_done_callback(self._connect_cb)
+        def callback():
+            self._task_connect = asyncio.ensure_future(
+                asyncio.open_connection(host, port),
+                loop=self._loop
+            )
+            self._grpc_connect_cb = grpc_connect_cb
+            self._task_connect.add_done_callback(self._connect_cb)
+
+        self._loop.call_soon_threadsafe(callback)
```
Comment (lidizheng) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/socket.pyx.pxi]: What's the performance impact? Alternatively, we can have a wrapper that handles the in-thread scheduling or cross-thread scheduling.

File: src/python/grpcio_tests/setup.py [hunk 9]
```diff
@@ -59,6 +59,7 @@
     'test_lite': commands.TestLite,
     'test_gevent': commands.TestGevent,
     'test_aio': commands.TestAio,
+    'test_asyncio': commands.TestAsyncio,
```
Comment (lidizheng) [inline src/python/grpcio_tests/setup.py]: nit: should we use a informative name that distinguish with `test_aio`? For example, `test_sync_over_aio`, `test_lite_over_aio`, `test_wrapped_aio`?

File: src/python/grpcio/grpc/_cython/_cygrpc/grpc.pxi [hunk 10]
```diff
@@ -26,6 +26,16 @@ ctypedef unsigned int       uint32_t
 ctypedef unsigned long long uint64_t
 
 
+cdef extern from "grpc/support/sync.h":
```
[the remark this answers is not in the document]
Comment (lidizheng) [reply src/python/grpcio/grpc/_cython/_cygrpc/grpc.pxi]: (I thought the Ruby build issue was blocking the usage of `absl/synchronization`)

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        self._loop_started = False
+
+        self._thread = threading.Thread(target=self._run_forever, daemon=True)
```
[the remark this answers is not in the document]
Comment (pfreixes) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: Ive updated the PR with an explanation of the performance for each of the different ways of running the stack comparing this PR and master.

TL;DR the performance penalty seems negligible

_stack_ | Master | PR
--- | --- | ---
sync | 11K | 11K
sync over Aio | x | 6K
Aio| 19K| 18K

More tests would need to be conducted with other arities for double-checking that effectively the footprint of adding this running thread is almost negligible.

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi (hunk 2 above)
Comment (pfreixes) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: Yes, I will migrate this for using `Event`. Thanks!

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+
+    def _loop_started_cb(self):
+        with self._loop_started_cv:
+            self._loop_started = True
+            self._loop_started_cv.notify_all()
+
+    def _run_forever(self):
+        self._asyncio_loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(self._asyncio_loop)
```
[the remark this answers is not in the document]
Comment (pfreixes) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: Nops, third-party event loops install their policy [1] which would make Asyncio aware which is the factory for creating new loops.

[1] https://github.com/pfreixes/grpcio_stress/blob/master/sync_unary.py#L113

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi (hunk 5 above)
Comment (pfreixes) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: Theoretically, if I'm not missing something I would say no. Exceptions are still raised through the futures. So, any exception happening in the IO loop that is bound to a future that is consumed in another thread will get that exception.

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi (hunk 1 above)
Comment (pfreixes) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: Yes Ive just used the `threading.Event` as being a good simplification of the `threading.Condition` where we do not take care of any variable.

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/socket.pyx.pxi (hunk 8 above)
Comment (pfreixes) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/socket.pyx.pxi]: Almost negligible, more info here [1]

[1] https://github.com/grpc/grpc/pull/22076/files#r381595041

File: src/python/grpcio/grpc/_cython/_cygrpc/grpc.pxi (hunk 10 above)
[the remark this answers is not in the document]
Comment (pfreixes) [reply src/python/grpcio/grpc/_cython/_cygrpc/grpc.pxi]: sorry some leftovers of an attempt for having a solution that would not require to acquire the GIL, this is not currently used in the proposed solution. I will remove it.

File: src/python/grpcio_tests/commands.py [hunk 13]
```diff
@@ -229,6 +229,52 @@ def run(self):
         if not result.value.wasSuccessful():
             sys.exit('Test failure')
 
+class TestAsyncio(setuptools.Command):
+    """Command to run tests w/asyncio."""
+
+    BANNED_TESTS = (
```
[the remark this answers is not in the document]
Comment (pfreixes) [reply src/python/grpcio_tests/commands.py]: Just followed the same structure as other `Test???` classes - there are many, I would prefer to follow the same strategy which is IMO super explicit.

File: src/python/grpcio_tests/commands.py [hunk 14]
```diff
@@ -229,6 +229,52 @@ def run(self):
...
+
+    def finalize_options(self):
+        # distutils requires this override.
+        pass
+
+    def run(self):
+        import grpc.experimental.aio
+        grpc.experimental.aio.init_grpc_aio()
+
+        import tests
```
[the remark this answers is not in the document]
Comment (pfreixes) [reply src/python/grpcio_tests/commands.py]: I would prefer for now use its own test with some bits duplicated, I like the idea for now for having this small duplication, which IMO gives more readability.

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+            # blocking the Asyncio loop.
+            return
+
+        if timeout_ms > 0:
+            self._io_ev.wait(timeout_ms * _RESOLUTION_MS)
+
+    cdef object asyncio_loop(self):
+        return self._asyncio_loop
+
+cdef _IOLoop _current_io_loop():
```
[the remark this answers is not in the document]
Comment (pfreixes) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: Unfortunately the `classmethod` is not yet available in Cython, so everything would need to be done by using a `staticmethod`. The reason for moving this function as an individual method was for making it usable in its C format from any Python object.

For using the `__new__` we would need to use a Python class If I'm not wrong.

If you believe that having a more clear encapsulation is a good trade-off happy to change it.

**Comment by pfreixes** [comment]
Found ou that once Ive switched to Python 3.6 some tests were failing randomly with this traceback:

```
Errors/Failures:
[91munit.call_test.TestStreamStreamCall.test_error_in_async_generator[0m
[1mtraceback:[0m
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/asyncio/base_tasks.py", line 27, in _task_get_stack
    f = task._coro.cr_frame
AttributeError: 'async_generator_athrow' object has no attribute 'cr_frame'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 608, in run
    self.tearDown()
  File "/Volumes/BuildData/tmpfs/src/github/grpc/workspace_python_macos_opt_asyncio/src/python/grpcio_tests/tests_aio/unit/_test_base.py", line 31, in wrapper
    return loop.run_until_complete(f(*args, **kwargs))
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/asyncio/base_events.py", line 468, in run_until_complete
    return future.result()
  File "/Volumes/BuildData/tmpfs/src/github/grpc/workspace_python_macos_opt_asyncio/src/python/grpcio_tests/tests_aio/unit/call_test.py", line 46, in tearDown
    await self._channel.close()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/asyncio/coroutines.py", line 110, in __next__
    return self.gen.send(None)
  File "/Volumes/BuildData/tmpfs/src/github/grpc/workspace_python_macos_opt_asyncio/py36_asyncio/lib/python3.6/site-packages/grpc/experimental/aio/_channel.py", line 295, in close
    await self._close(grace)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/asyncio/coroutines.py", line 110, in __next__
    return self.gen.send(None)
  File "/Volumes/BuildData/tmpfs/src/github/grpc/workspace_python_macos_opt_asyncio/py36_asyncio/lib/python3.6/site-packages/grpc/experimental/aio/_channel.py", line 255, in _close
    stack = task.get_stack(limit=1)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/asyncio/base_tasks.py", line 29, in _task_get_stack
    f = task._coro.gi_frame
AttributeError: 'async_generator_athrow' object has no attribute 'gi_frame'

[1mstdout:[0m
None
[1mstderr:[0m
None
```

Ive tried to reproduce in master but I couldn't, so considering that with Python 3.7 the branch was passing all of the tests most likely it's something related to the changes proposed and the way that the channel closes the ongoing calls.

Any idea will help /cc @lidizheng

File: src/python/grpcio_tests/commands.py [hunk 16]
```diff
@@ -229,6 +229,52 @@ def run(self):
...
+
+    def initialize_options(self):
+        pass
+
+    def finalize_options(self):
+        # distutils requires this override.
+        pass
+
+    def run(self):
+        import grpc.experimental.aio
```
[the remark this answers is not in the document]
Comment (pfreixes) [reply src/python/grpcio_tests/commands.py]: Updated the docstring of the Class and also changed the name of the class for a more meaningful name.

File: src/python/grpcio_tests/setup.py (hunk 9 above)
Comment (pfreixes) [reply src/python/grpcio_tests/setup.py]: renamed to `test_sync_over_aio`

**Comment by lidizheng** [comment]
@pfreixes This failure looks deterministic to me, looks like a bug in CPython. I'll investigate if we can work around it.

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+        self._loop_started_ev.set()
+
+    def _run_forever(self):
+        self._asyncio_loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(self._asyncio_loop)
+        self._asyncio_loop.call_soon(self._loop_started_cb)
+        try:
+            self._asyncio_loop.run_forever()
+        except Exception as exp:
+            sys.stderr("An error ocurred with the IO loop {}{}".format(exp, os.linesep))
```
Comment (gnossen) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: `sys.stderr.write`?

Comment (lidizheng) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: I missed this one. Should we use `_LOGGER.error` or `_LOGGER.exception`?

Comment (pfreixes) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: facepalm!

Ive used the `sys.stderr` because for an unknown reason `print` within Cyhton does not have support for the `file` kw, no idea the reason why.

I can use the `_LOGGER.exception` as it's suggested by Lidi. do you agree?

Comment (gnossen) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: Yeah. `_LOGGER` is probably better.

**Comment by lidizheng** [comment]
@pfreixes The frame object not found issue seems like a bug in CPython. In short, there is a global callback for closing async generators. It schedules an `aclose` task onto the event loop, however, `aclose` task is a `PyAsyncGenAThrow` in CPython which doesn't have proper frame attributes. Then it fails.

```python
            try:
                stack = task.get_stack(limit=1)
            except AttributeError as attribute_error:
                # NOTE(lidiz) tl;dr: If the Task is created with a CPython
                # object, it will trigger AttributeError.
                #
                # In the global finalizer, the event loop schedules
                # a CPython PyAsyncGenAThrow object.
                # https://github.com/python/cpython/blob/00e45877e33d32bb61aa13a2033e3bba370bda4d/Lib/asyncio/base_events.py#L484
                #
                # However, the PyAsyncGenAThrow object is written in C and
                # failed to include the normal Python frame objects. Hence,
                # this exception is a false negative, and it is safe to ignore
                # the failure (someone should fix it in CPython!).
                # https://github.com/python/cpython/blob/a025d4ca99fb4c652465368e0b4eb03cf4b316b9/Objects/genobject.c#L1989
                if 'frame' in str(attribute_error):
                    continue
                else:
                    raise
```

Even with this update, I still observe deadlocks occasionally. More investigation is needed.

**Comment by pfreixes** [comment]
@lidizheng definitely seems a bug from CPython, and it's not only happening with the `async_generator_throw`, but also happens with all of the versions of the `async_generator_*`. Indeed the only one that supports stack introspection is `async_generator` and has its own attribute named `ag_frame` [1].

So technically with the fix that you proposed it should work, I will start by using the patch that you provided and see where these deadlocks are happening.

Eventually, we would need to send a patch to CPython for fixing this mess. Since you found the bug, I think that you should do it.

[1] https://github.com/python/cpython/blob/master/Objects/genobject.c#L1329

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+_SECONDS_PER_MILLISECOND = 0.001
+cdef _IOLoop _io_loop = None
+
+cdef class _IOLoop:
+
+    def __init__(self):
+        global _io_loop
+
+        # Only one instantiation is exepected        
+        assert _io_loop is None
```
Comment (rmariano) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: I'd add something similar to what the comment says to the assertion message, to make it more understanable for developers
```suggestion
        assert _io_loop is None, "More than one loop found"
```

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi (hunk 17 above)
Comment (rmariano) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/io_loop.pyx.pxi]: +1 to using ``_LOGGER``

**Comment by lidizheng** [comment]
@pfreixes CPython PR filed https://github.com/python/cpython/pull/18669. No idea how long it will take.

**Comment by pfreixes** [comment]
@lidizheng I have bad news, once we solve the deadlocks by releasing the GIL - last pushed commits - I'm experimenting the following two down-side effects:

- We lose ~15% of performance, moving from 19K QPS to 15QPS!
- Im not able to have all `test_sync_over_aio` in green again. Something that I got before the changes that I pushed Today

Any ideas about what could be the way to go will help, because I'm starting to feel that maybe this shortcut will be never fully functional, and having as a side-effect hurting the Aio module, as it is being reflected with the performance.

**Comment by pfreixes** [comment]
My next plans for this PR:

- Work on having the legacy tests passing in green as we had before fixing the deadlocks
- Measure the performance impact on the server-side
- Decision point with you once we have the tests passing and we have numbers of the regression in the server-side and the client-side.

**Comment by pfreixes** [comment]
CI is green, good news. But there is still a lot of work to do. I will continue with the plan. I will dismiss your review @gnossen since there is a lot of work to do, and the performance after adding the proper GIL lock/unlock methods has been impacted

**Comment by pfreixes** [comment]
With the latest changes for preventing deadlocks, we added a substantial performance regression in the `Aio` package. This regression has paired - for bad - the performance of the asynchronous driver and the synchronous one. Benchmarks showed that both of them were able to reach a maximum of 12K QPS, when in the master branch the asynchronous one is able to reach 20K QPS./

**Why this regression appeared?**

The main reason for releasing/acquiring the GIL, some examples here [1] [2], when we called gRPC Core functions and when we were called by gRPC Core functions was for addressing the deadlocks that appeared at the moment that we implemented and offload all IO operations to an isolated IO loop.

The deadlocks appeared between gRPC Core mutex and the GIL. Without releasing or acquiring the GIL explicitly by us this what was happening.

1 - (thread 1) (Pyton) call a gRPC Core function
2 - (thread 1) (gRPC core) create a mutex X
3 - (thread 1) (gRPC core) calls on_callback
4 - (thread 1) (Python) syncrhonized the thread and releases the GIL implicitly
5 - (thread 2) (Python) call a gRPC Core function
6 - (thread 2) (gRPC core) tried to acquire the mutex X

At that point, thread 1 and thread 2 had a deadlock because the mutex X was already locked by thread 1 and the GIL was locked by thread 2. This didn't happen before because the execution of the `on_callback` would never release the GIL implicitly and would return the control back to the gRPC core function which would eventually release the mutex X.

For addressing the situation we had to release explicitly the GIL every time that we would make a call to gRPC core, so previous flow would look like:

1 - (thread 1) (Pyton) call a gRPC Core function
2 - (thread 1) (Pyton) Release the GIL
3 - (thread 1) (gRPC core) create a mutex X
4 - (thread 1) (gRPC core) calls on_callback
5 - (thread 1) (Python) Acquire the GIL
6 - (thread 1) (Python) synchronizes the thread and RELEASE the GIL implicitly
7 - (thread 2) (Python) Acquire the GIL
8 - (thread 2) (Python) call a gRPC Core function
9 - (thread 2) (Python) Release the GIL
10 - (thread 2) (gRPC core) tried to acquire the mutex X
11 - (thread 1) (Python) Acuire the GIL
12 - ...

As can bee observed, by acquiring and releasing the GIL we prevented the deadlock scenario where a lock from the gRPC Core and the GIL might end up creating a deadlock.

But addressing this didn't come for free, the following is the profiling of the top15 functions for that specific branch vs the top15 form master.

Current branch
```bash
Total: 3740 samples
     500  13.4%  13.4%      500  13.4% __pthread_cond_signal
     398  10.6%  24.0%      494  13.2% __pthread_cond_timedwait
     161   4.3%  28.3%     3739 100.0% _PyEval_EvalFrameDefault
     128   3.4%  31.7%      128   3.4% syscall
     114   3.0%  34.8%      114   3.0% PySequence_Contains
     111   3.0%  37.8%      111   3.0% __lll_unlock_wake
      75   2.0%  39.8%      180   4.8% _PyType_Lookup
      60   1.6%  41.4%       61   1.6% __libc_write
      57   1.5%  42.9%       57   1.5% __libc_read
      48   1.3%  44.2%      339   9.1% PyObject_GC_UnTrack
      47   1.3%  45.4%      209   5.6% _PyObject_GenericGetAttrWithDict
      40   1.1%  46.5%       42   1.1% __GI_epoll_pwait
      38   1.0%  47.5%       38   1.0% PyObject_Malloc
      37   1.0%  48.5%     3129  83.7% _init@36e8
      33   0.9%  49.4%       97   2.6% _PyDict_GetItem_KnownHash
```

master
```
(pprof) top15
Total: 3138 samples
     181   5.8%   5.8%      181   5.8% syscall
     174   5.5%  11.3%     3137 100.0% _PyEval_EvalFrameDefault
     170   5.4%  16.7%      170   5.4% __pthread_cond_signal
     139   4.4%  21.2%      141   4.5% PySequence_Contains
     113   3.6%  24.8%      147   4.7% __pthread_cond_timedwait
      74   2.4%  27.1%       74   2.4% __libc_write
      65   2.1%  29.2%       69   2.2% __GI_epoll_pwait
      60   1.9%  31.1%      175   5.6% _PyType_Lookup
      56   1.8%  32.9%       56   1.8% PyObject_Malloc
      53   1.7%  34.6%      235   7.5% PyObject_GC_UnTrack
      51   1.6%  36.2%      192   6.1% _PyObject_GenericGetAttrWithDict
      47   1.5%  37.7%       49   1.6% __libc_read
      37   1.2%  38.9%     2408  76.7% _init@36e8
      37   1.2%  40.1%       37   1.2% _init@60860
      29   0.9%  41.0%       29   0.9% PyType_IsSubtype
```

As we can see the two offending functions `__pthread_cond_signal` and `__pthread_cond_timedwait` are involved in GIL acquiring [3] and releasing had a substantial increase in the total number of times called and the time spent on them.

Before we introduced the changes for acquiring and releasing the GIL properly - for addressing the deadlock issues - the profile showed that the time spent by the two offending functions was only slightly higher - hence, the little regression introduced and claimed initially.

[1] https://github.com/grpc/grpc/pull/22076/files#diff-cdd81de112d1cded21f7544d8264b12aR81
[2] https://github.com/grpc/grpc/pull/22076/files#diff-cdd81de112d1cded21f7544d8264b12aR114
[3] https://github.com/python/cpython/blob/e225bebc1409bcf68db74a35ed3c31222883bf8f/Python/ceval_gil.h#L184

**Comment by pfreixes** [comment]
Closing this in favor of these two ones https://github.com/grpc/grpc/pull/22258 https://github.com/grpc/grpc/pull/22259
